staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -506,6 +510,101 @@ func TestCacheDontAcceptRequestsStopped(t *testing.T) {
 	}
 }
 
+func TestCacherDontMissEventsOnReinitialization(t *testing.T) {
```

Comment:
This newly added test is fairly reliably reproducing the problem without the actual fix (although requires the change that I did switching to `time.Nanosecond`  in the production code - which fwiw it's useful anyway).

The failure manifests with the following log:
```
    cacher_whitebox_test.go:597: Unexpected event received, prevRV=7, rv=12
```

and can be easily achieved using:
```
make test WHAT='k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher -run=MissEvents -count=10'
```

With the fix, this test is reliably passing 200 tests multiple times.
```
make test WHAT='k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher -run=MissEvents -count=200'
```

but for now the test is sometimes still failing with
```
    cacher_whitebox_test.go:597: Unexpected event received, prevRV=12, rv=3
```

This isn't a race - it seems like a different problem with init events which are returned in random order - but I need to confirm it.

Reply:
OK - I understand the second problem now.

Unfortunately that's another bug - that again isn't highly to happen, but as shown above actually can happen.
The fix for it should be simple, but I'm going to look into it tomorrow/early next week.